build_tools/cmake/iree_bytecode_module.cmake

```diff
@@ -114,6 +114,23 @@ function(iree_bytecode_module)
     list(APPEND _ARGS "-iree-llvm-embedded-linker-path=\"${_EMBEDDED_LINKER_TOOL_EXECUTABLE}\"")
   endif()
 
+  # Support testing in TSan build dirs. Unlike other sanitizers, TSan is an
+  # ABI break: when the host code is built with TSan, the module must be too,
+  # otherwise we get crashes calling module code.
+  #
+  # We make an exception for cross-compiling tests, crudely identified for now
+  # by the presence of a target-triple flag. IREE_ENABLE_TSAN implies (for the
+  # time being) a host build dir, so any such tests are by definition build-only
+  # which makes it unnecessary to enable TSan in module code. We could support
+  # that if needed, it's just that for now, as TSan linking is implemented using
+  # clang as the linker driver, we would have to figure out the right (-march?)
+  # flag to pass for cross-linking.
+  if (IREE_ENABLE_TSAN AND NOT (_ARGS MATCHES "-iree-llvm-target-triple="))
+    list(APPEND _ARGS "-iree-llvm-sanitize=thread")
+    # Avoid undefined-symbol errors on tsan symbols e.g. __tsan_read8
+    list(APPEND _ARGS "-iree-llvm-link-embedded=false")
+  endif()
```

Comment:
ooo today but this is the high level question that i'd need to understand better before going into impl details:  My goal here has been "Building and running tests (CTest) in a build dir with `IREE_ENABLE_TSAN=ON` should work".  Are you (1) agreeing with the goal but suggesting an alternate impl strategy or (2) suggesting an alternate goal?  (Either way please elaborate on the alternate suggestion, more than happy to implement whatever reviewers suggest - I just need something actionable).

Reply:
Having a single setting that affects all builds/tests is ambitious, so I'd start more focused, based on priority and expected usage. Is there a particular part of the project you want TSAN coverage for?

1. Base runtime tests/binaries (cc_test, some of the HAL CTS, tools like generate_embed_data)
2. Compiler-adjacent runtime tests/binaries (check_test/iree-run-module not using CPU codegen)
3. Compiler tests/binaries (lit_test, iree-opt)
4. CPU compiled code (check_test/iree-run-module using LLVM CPU codegen)

It looks like you're jumping straight to 4 (the most complex), when excluding CPU generated code from TSAN would be a simpler way to start. We can exclude LLVM CPU codegen tests from a full "build+run all tests" config if we mainly want coverage for code outside of CPU codegen, for example by only enabling `IREE_TARGET_BACKEND_VMVX` and `IREE_HAL_DRIVER_VMVX`.

If we really do want coverage for the cpu codegen, then it's a question of whether we need to fully integrate it into the build system as you're close to here, or if just keeping support around for offline developer use is sufficient. The `iree-llvm-sanitize` flag has been in that state (not wired up to any CI) since it was added.

Reply:
I really have this high level goal that one should be able to toggle `IREE_ENABLE_TSAN=ON`, rebuild, rerun tests, and that would do what one would naively think it does: the same tests are now running with TSan enabled. If I followed the approach outline in the previous comment, only enabling the VMVX driver/backend, then setting `IREE_ENABLE_TSAN=ON` would either (the current behavior) cause CPU codegen tests to fail, or (the only alternative) be skipped. We would be digging deeper the hole that we are already in with the complex matrix of cmake options that work / don't work with each other, and the complex set of tests that are expected to fail in each case.   I'm interested in fixing this high level problem that the IREE project currently has.